stdlib/json.gr

```diff
@@ -2069,3 +2069,298 @@ provide let parse: (str: String) => Result<Json, JsonParseError> = (str: String)
+   * 
+   * @param lens: The lens to apply to the subject data
+   * @param fn: The function to apply to the matched data at the lens if matched
+   * @param subject: The data which will have the lens applied onto it
+   * @returns The new data after the lens mapping has been applied wrapped in `Some` if the lens matches the given data, or `None` if the data cannot be matched to the lens
+   * 
+   * @example assert map(number, x => x * 2, JsonNumber(5)) == Some(JsonNumber(10))
+   * @example
+   * assert map(prop("x"), x => JsonArray([x, x]), JsonObject([("x", JsonNumber(1))])) ==
+   *   Some(JsonObject([("x", JsonArray([JsonNumber(1), JsonNumber(1)]))]))
+   * @example assert map(number, x => x * 2, JsonString("abc")) == None
+   */
+  provide let map = (lens, fn, subject) => {
+    match (lens.get(subject)) {
+      Some(lensVal) => lens.set(fn(lensVal), subject),
+      None => None,
+    }
+  }
+
+  /**
+   * A lens whose focus is a JSON value.
+   * 
+   * @example assert get(json, JsonString("abc")) == Some(JsonString("abc"))
+   * 
+   * @since v0.7.0
+   */
+  provide let json = {
+    get: json => Some(json),
+    set: (newValue, _) => Some(newValue),
+  }
+
+  /**
+   * A lens whose focus is a JSON boolean value.
+   * 
+   * @example assert get(boolean, JsonBoolean(true)) == Some(true)
+   * 
+   * @since v0.7.0
+   */
+  provide let boolean = {
+    get: json => {
+      match (json) {
+        JsonBoolean(val) => Some(val),
+        _ => None,
+      }
+    },
+    set: (newValue, _) => Some(JsonBoolean(newValue)),
+  }
+
+  /**
+   * A lens whose focus is a JSON string value.
+   * 
+   * @example assert get(string, JsonString("abc")) == Some("abc")
+   * 
+   * @since v0.7.0
+   */
+  provide let string = {
+    get: json => {
+      match (json) {
+        JsonString(val) => Some(val),
+        _ => None,
+      }
+    },
+    set: (newValue, _) => Some(JsonString(newValue)),
+  }
+
+  /**
+   * A lens whose focus is a JSON number value.
+   * 
+   * @example assert get(number, JsonNumber(123)) == Some(123)
+   * 
+   * @since v0.7.0
+   */
+  provide let number = {
+    get: json => {
+      match (json) {
+        JsonNumber(val) => Some(val),
+        _ => None,
+      }
+    },
+    set: (newValue, _) => Some(JsonNumber(newValue)),
+  }
+
+  /**
+   * A lens whose focus is a JSON array.
+   * 
+   * @example assert get(array, JsonArray([JsonNumber(123)])) == Some([JsonNumber(123)])
+   * 
+   * @since v0.7.0
+   */
+  provide let array = {
+    get: json => {
+      match (json) {
+        JsonArray(val) => Some(val),
+        _ => None,
+      }
+    },
+    set: (newValue, _) => Some(JsonArray(newValue)),
+  }
+
+  /**
+   * A lens whose focus is the properties of a JSON object.
+   * 
+   * @example assert get(objectProps, JsonObject([("a", JsonNumber(123))])) == Some([("a", JsonNumber(123))])
+   * 
+   * @since v0.7.0
+   */
+  provide let objectProps = {
```

Comment:
I feel like `objectProperties` might be a bit verbose but we have been pretty against shorthands so far thoughts? same as for `prop`